### PR TITLE
[KYUUBI #2104] Kill yarn job using yarn client API when kyuubi engine …

### PR DESCRIPTION
**Pull request by 942011334**
…initialization times out and yarn application status is accepted

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/pom.xml [hunk 1]
```diff
@@ -385,6 +385,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+            <version>3.4.6</version>
```
Comment (yaooqinn) [inline kyuubi-server/pom.xml:392]: can we add it to dependencyManagement then inherit it here？

**Comment by yaooqinn** [comment]
thanks @942011334  for the contibution, also cc @wForget @SteNicholas @pan3793

**Comment by 942011334** [comment]
good idea .




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Kent ***@***.***&gt;; 
发送时间: 2022年3月12日(星期六) 下午4:34
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/incubator-kyuubi] [Kyuubi#2104] Kill yarn job using yarn client API when kyuubi engine … (PR #2110)





 
@yaooqinn commented on this pull request.
 
 
In kyuubi-server/pom.xml:
 &gt; @@ -385,6 +385,13 @@              <scope&gt;test</scope&gt;          </dependency&gt;   +        <dependency&gt; +            <groupId&gt;org.mockito</groupId&gt; +            <artifactId&gt;mockito-core</artifactId&gt; +            <scope&gt;test</scope&gt; +            <version&gt;3.4.6</version&gt;  
can we add it to dependencyManagement then inherit it here？
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: kyuubi-server/pom.xml (hunk 1 above)
Comment (942011334) [reply kyuubi-server/pom.xml:392]: ok, good idea.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 2]
```diff
@@ -44,6 +45,8 @@ class SparkProcessBuilder(
 
   import SparkProcessBuilder._
 
+  var yarnClient = YarnClient.createYarnClient
```
Comment (SteNicholas) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:48]:
```suggestion
  val yarnClient = getYarnClient
  
  def getYarnClient: YarnClient = YarnClient.createYarnClient
```
In the test, you could override the `getYarnClient` method.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 3]
```diff
@@ -195,20 +198,20 @@ class SparkProcessBuilder(
...
+        try {
+          val hadoopConf = KyuubiHadoopUtils.newHadoopConf(conf)
+          yarnClient.init(hadoopConf)
+          yarnClient.start()
+          val applicationId = ApplicationId.fromString(appId)
+          yarnClient.killApplication(applicationId)
+          s"Killed Application $appId successfully."
+        } catch {
+          case e: Throwable =>
+            s"Failed to kill Application $appId, Caused by ${e.getMessage}."
```
Comment (SteNicholas) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:210]: Adds the `Please kill it manually.` in the message.

**Comment by SteNicholas** [comment]
Please help to update the title to `[Kyuubi #2104] Kill yarn job using yarn client API when kyuubi engine …`.

File: kyuubi-server/pom.xml (hunk 1 above)
Comment (pan3793) [reply kyuubi-server/pom.xml:392]: Please use scalatestplus-mockito instead.
```
<properties>
    <scalatestplus.version>3.2.9.0</scalatestplus.version>
</properties>

<dependency>
    <groupId>org.scalatestplus</groupId>
    <artifactId>mockito-3-4_${scala.binary.version}</artifactId>
    <version>${scalatestplus.version}</version>
</dependency>
```

File: pom.xml [hunk 4]
```diff
@@ -127,6 +127,7 @@
         <kubernetes-client.version>5.12.1</kubernetes-client.version>
         <ldapsdk.version>5.1.4</ldapsdk.version>
         <mysql.jdbc.version>8.0.27</mysql.jdbc.version>
+        <scalatestplus.version>3.2.9.0</scalatestplus.version>
```
Comment (pan3793) [inline pom.xml:130]: Move this line after `<scalatest.version>`

File: kyuubi-server/pom.xml [hunk 5]
```diff
@@ -386,10 +386,8 @@
         </dependency>
 
         <dependency>
-            <groupId>org.mockito</groupId>
-            <artifactId>mockito-core</artifactId>
-            <scope>test</scope>
-            <version>3.4.6</version>
+            <groupId>org.scalatestplus</groupId>
+            <artifactId>mockito-3-4_${scala.binary.version}</artifactId>
```
Comment (pan3793) [inline kyuubi-server/pom.xml:390]: `<scope>test</scope>`

**Comment by pan3793** [comment]
cc @cxzl25

**Comment by 942011334** [comment]
hello , when&nbsp;use scalatestplus-mockito instead ,"build/dependency.sh" test failed, help me to see the problem , tanks!




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Kent ***@***.***&gt;; 
发送时间: 2022年3月12日(星期六) 晚上9:14
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/incubator-kyuubi] [KYUUBI #2104] Kill yarn job using yarn client API when kyuubi engine … (PR #2110)





 
@pan3793 commented on this pull request.
 
 
In kyuubi-server/pom.xml:
 &gt; @@ -385,6 +385,13 @@              <scope&gt;test</scope&gt;          </dependency&gt;   +        <dependency&gt; +            <groupId&gt;org.mockito</groupId&gt; +            <artifactId&gt;mockito-core</artifactId&gt; +            <scope&gt;test</scope&gt; +            <version&gt;3.4.6</version&gt;  
Please use scalatestplus-mockito instead.
 <properties&gt;     <scalatestplus.version&gt;3.2.9.0</scalatestplus.version&gt; </properties&gt; <dependency&gt;     <groupId&gt;org.scalatestplus</groupId&gt;     <artifactId&gt;mockito-3-4_${scala.binary.version}</artifactId&gt;     <version&gt;${scalatestplus.version}</version&gt; </dependency&gt;  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by pan3793** [comment]
> hello , when&nbsp;use scalatestplus-mockito instead ,"build/dependency.sh" test failed, help me to see the problem , tanks!
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "Kent ***@***.***&gt;; 发送时间: 2022年3月12日(星期六) 晚上9:14 收件人: ***@***.***&gt;; 抄送: ***@***.***&gt;; ***@***.***&gt;; 主题: Re: [apache/incubator-kyuubi] [KYUUBI #2104] Kill yarn job using yarn client API when kyuubi engine … (PR #2110) @pan3793 commented on this pull request. In kyuubi-server/pom.xml: &gt; @@ -385,6 +385,13 @@ <scope&gt;test</scope&gt; </dependency&gt; + <dependency&gt; + <groupId&gt;org.mockito</groupId&gt; + <artifactId&gt;mockito-core</artifactId&gt; + <scope&gt;test</scope&gt; + <version&gt;3.4.6</version&gt; Please use scalatestplus-mockito instead. <properties&gt; <scalatestplus.version&gt;3.2.9.0</scalatestplus.version&gt; </properties&gt; <dependency&gt; <groupId&gt;org.scalatestplus</groupId&gt; <artifactId&gt;mockito-3-4_${scala.binary.version}</artifactId&gt; <version&gt;${scalatestplus.version}</version&gt; </dependency&gt; — Reply to this email directly, view it on GitHub, or unsubscribe. Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

You pinned `<artifactId>mockito-3-4_${scala.binary.version}</artifactId>` in `compile` scope, please take a look at my comments

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala [hunk 6]
```diff
@@ -242,7 +244,7 @@ class SparkProcessBuilderSuite extends KerberizedTestHelper {
   test("kill application") {
     val pb1 = new FakeSparkProcessBuilder(conf) {
       override protected def env: Map[String, String] = Map()
-      override def getYarnClient: YarnClient = mock(classOf[YarnClient])
+      override def getYarnClient: YarnClient = mock(classTag[YarnClient])
```
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilderSuite.scala:247]: `mock(classTag[YarnClient])` can be simplified to `mock[YarnClient]`

File: pom.xml [hunk 7]
```diff
@@ -888,6 +888,13 @@
                 <scope>test</scope>
             </dependency>
 
+            <dependency>
+                <groupId>org.scalatestplus</groupId>
+                <artifactId>mockito-3-4_${scala.binary.version}</artifactId>
+                <version>${scalatestplus.version}</version>
+                <scope>test</scope>
```
Comment (pan3793) [inline pom.xml:895]: we don't need `<scope>test</scope>` here

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 8]
```diff
@@ -44,6 +45,10 @@ class SparkProcessBuilder(
 
   import SparkProcessBuilder._
 
+  val yarnClient = getYarnClient
+
+  def getYarnClient: YarnClient = YarnClient.createYarnClient
```
Comment (wForget) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:50]: The name of `getYarnClient` method seems to be a little confused. It looks like the `getter` method of `yarnClient` variable.

**Comment by 942011334** [comment]
yes, the purpose is to override this method for the test class .




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Kent ***@***.***&gt;; 
发送时间: 2022年3月13日(星期天) 中午1:20
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/incubator-kyuubi] [KYUUBI #2104] Kill yarn job using yarn client API when kyuubi engine … (PR #2110)





 
@wForget commented on this pull request.
 
 
In kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:
 &gt; @@ -44,6 +45,10 @@ class SparkProcessBuilder(      import SparkProcessBuilder._   +  val yarnClient = getYarnClient + +  def getYarnClient: YarnClient = YarnClient.createYarnClient  
The name of getYarnClient method seems to be a little confused. It looks like the getter method of yarnClient variable.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala (hunk 8 above)
Comment (SteNicholas) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:50]: @wForget, what about `createYarnClient`? @942011334, sorry to suggest a bit confusing method name.

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:50]: > what about createYarnClient?

LGTM

**Comment by 942011334** [comment]
Thanks to every expert's guidance, support and help.






------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Kent ***@***.***&gt;; 
发送时间: 2022年3月13日(星期天) 下午2:12
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/incubator-kyuubi] [KYUUBI #2104] Kill yarn job using yarn client API when kyuubi engine … (PR #2110)





 
@wForget commented on this pull request.
 
 
In kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:
 &gt; @@ -44,6 +45,10 @@ class SparkProcessBuilder(      import SparkProcessBuilder._   +  val yarnClient = getYarnClient + +  def getYarnClient: YarnClient = YarnClient.createYarnClient   
what about createYarnClient?
  
LGTM
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by yaooqinn** [comment]
thanks @942011334 for the first contribution, merged to master

**Pull request closed without merging**